File: 23-DashInsert/23.js
//Given a number, output a string that prints a dash next to every two odd numbers

function DashInsert(str) { 
  return (str + "").split("").reduce(function(acc,x,i,arr){
    return acc + ((parseInt(x)%2 === 1 && parseInt(arr[i-1])%2 === 1 ) ? '-' + x : x);
  },"");
}

function DashInsertRegex(str){ //regex, yay~
  return (str + "").replace(/([13579])(?=([13579]))/g, '$1-');
}

console.log(DashInsert(99999)  === "9-9-9-9-9");
console.log(DashInsert(2468)   === "2468");
console.log(DashInsert(454793) === "4547-9-3");
console.log(DashInsertRegex(99999)  === "9-9-9-9-9");
console.log(DashInsertRegex(2468)   === "2468");
console.log(DashInsertRegex(454793) === "4547-9-3");

File: 23-DashInsert/23.hs
--Given a number, output a string that prints a dash next to every two odd numbers
import Data.Char
import Text.Regex

dashInsert :: Int -> String
dashInsert num = splitOdds $ show num 
  where splitOdds (x:y:zs)
          | odd (digitToInt x) && odd (digitToInt y) = x:'-':splitOdds(y:zs)
          | otherwise                                = x:splitOdds(y:zs)
        splitOdds xs = xs

main = do
  print $ dashInsert 99999  == "9-9-9-9-9"
  print $ dashInsert 2468   == "2468"
  print $ dashInsert 454793 == "4547-9-3"